### ISSUE TITLE
Need help to make result closer to original work 

### ISSUE THREAD

**Issue by gskachkov** [NickLucche/pyramid-cnn-leaves-segmentation#1]
Hi Nicolo,

Thanks for you implementation of the **A Pyramid CNN for Dense-Leaves Segmentation** in pyTorch. I've checked and it worked great!

Also I've run original implementation. 
I have small question, what do you think  might be changes in python code to have result similar to result of original paper? Some of the leafs are united into one leaf. I'm quite newbie in Computer vision and deep learning :-(
For instance results for image leaf00025_000_img.png 
<img width="1261" alt="Screen Shot 2020-05-03 at 9 59 26 PM" src="https://user-images.githubusercontent.com/4946944/80923110-21c84e00-8d8a-11ea-89f6-2667e6d648bb.png">
Images:
1. Original image
2. Result from current repo
3. From original implementation

**Comment by NickLucche** [comment]
Hi Oleksander,
I'm glad the implementation can be of help.

What's the code you're running to produce the result from the original implementation?
The most fundamental difference between this repo and the author's work is the 'post-processing' part which is applied to the net's output.
Here you only have access to the raw network's output, while the paper also devises a method for refining and merging the segments composing the leaves' contours, entirely relying on well known traditional CV techniques (+ author's ideas) which I did not implement here since they diverge from the pure-ML approach I was after, tho it's a fundamental part of the process.

If you're indeed only comparing the results of the two nets, then the main difference might be in the (scarce) training of the network I provided.

Let me know and I'll see if I can provide you with a better model or give you some good pointers to try it out on your own! :)

**Comment by gskachkov** [comment]
Thanks for quick answer!

I've used matlab code that is provided by author https://github.com/dmorris0/PyramidCNN 
If I'm correctly understand following matlab code, my result is produced just by network and without 'post-processing'
https://github.com/dmorris0/PyramidCNN/blob/master/MatlabCode/scoreEdgeNetOnData.m#L45
https://github.com/dmorris0/PyramidCNN/blob/master/MatlabCode/runNetEdges.m#L105

Yeah you are right, 'post-processing' is used for making instance segmentation and there are used watershed segmentation. Really slow - because GPU is not used :( For instance result for already provided file  leaf00025_000_img.png
![leaf00025_000_img-segments](https://user-images.githubusercontent.com/4946944/80924459-b97d6a80-8d91-11ea-8004-0e05ff98b836.png)

It will be great if you can provide some tips, parameters or suggestion how to receive better model!

**Comment by NickLucche** [comment]
Yes I believe that if the output you reported is the one you get by running:
> `edgelist = runNetEdges( net, imglist, edgefolder, doGPU, redoEdges, crop, averageScales );`

then I guess it's simply a matter of network behaving better, which we can address here too.

In order of importance, you can try re-training the network by:
- Choosing 'correctly' the weights which are applied to each prediction at different scale using the arguments `loss_weights` when creating the model `model = PyramidNet(n_layers=5, loss_weights=[torch.tensor([1.0])]*5)`. 
This is usually pretty important for predictions with such a high imbalance in terms of class. I tried a few different values by estimating a rough rate of `num_edges_pixels/num_pixels` over the whole dataset, but I was unsuccessful in achieving correct results.

- Use data augmentation while training (my DataLoader still has a little bug regarding that which I want to fix as soon as I can re-train the whole model).

- Train for a longer amount of epochs wrt what I did (100+epochs) by checking the loss more carefully.

- The very last difference with the paper is that I did not re-weight the most "difficult" examples, the ones with highest loss, every 5 epochs.

Let me know if you need more advice!
Hopefully I'll also be able to re-train the whole model during this week

**Comment by gskachkov** [comment]
Thanks for your tips!
I'll try to follow them and see if I have better results. 

There are small question, not directly related to started one.

1. I've tried to play with different type of leafs and I see that it split complex leaf into several leafs.
What do you think If I retrain with grape leafs, will it fix, or it is better to use another approach or paper?
There is result from original source code:
source
![IMG-362_left](https://user-images.githubusercontent.com/4946944/81003896-3cff9000-8e54-11ea-98c5-08bc2926d5bc.png)
edges
![IMG-362_left_img-edge](https://user-images.githubusercontent.com/4946944/81003817-1e999480-8e54-11ea-910b-fd92a657f809.png)
segments
![IMG-362_left_img-segments](https://user-images.githubusercontent.com/4946944/81003827-235e4880-8e54-11ea-92b5-3f8bc3895380.png)

2. What do you think, will it be possible to rewrite the watershed segmentation to use the GPU and speedup execution of the segmentation, currently it very slow, for instance with using such libs https://github.com/louismullie/watershed-cuda or similar lib?
or it is better give away using the watershed and something completely different?

**Comment by NickLucche** [comment]
1. Yes definitely, if you manage to train with the specific type of leaves you're after you can expect a boost in performance for your application. But the problem is a difficult one, depending on how many leaves you need to have segmented exactly; also I'm not aware of many open datasets on the matter 😢 
2. Yes that's an excellent idea, especially if you can leverage some pre-existing CUDA implementation. Keep in mind that the author's implementation is built on the watershed algorithm by adding something to the pipeline; I'm pretty sure tho that there are different ways to approach this, especially if you want an optimal solution to your particular problem. AKA, you should experiment a bit to find the right setting!

Let me know if you have news!

**Comment by gskachkov** [comment]
Thanks for the feedback! I'll back as soon as have new results :-)

**Comment by NickLucche** [comment]
I re-trained the model achieving slightly better results, I'll see if I'll have availability next days to experiment some more.
I hope it provides a slightly better base to build upon in your application; here's a quick test on your image:

![custom_prediction](https://user-images.githubusercontent.com/10706289/81335074-50ef0000-90a7-11ea-9d7f-25da8beae0be.png)

**Comment by gskachkov** [comment]
Great! Looks like good progress, I can count more leaf outlines :-) Thanks!

**Comment by Ostyk** [comment]
I've actually recently also implemented this network in Pytorch but took a different approach. I used cross entropy -- softmax and forced it to predict 3 outputs (background, interior and edge) and weighting them as in the paper
`Pixel loss and weighting: A softmax loss is used to specify a cross entropy function on edge and interior pixels. Since the ground truth boundary pixels are signif- icantly outnumbered by leaf interior pixels, following [9], the boundary and interior pixels are reweighted to give them equal overall influence. Unlabeled pixels are given zero weight. `

These are the results
![Screenshot 2020-05-08 at 11 53 03](https://user-images.githubusercontent.com/15198671/81394737-b7fdca80-9122-11ea-8be8-e18fd5d6c5c9.png)

**Comment by NickLucche** [comment]
Hi Michael,
if I understood correctly you got rid of the loss masking and you used that instead to introduce a third class; I had the same idea but in the end I stuck to the paper approach, glad to see it worked so well tho!
Predictions look really on point.

Did you train on the same 128x128 resolution or you tried pushing it a bit farther?

**Comment by Ostyk** [comment]
Hi Nicolo,

I actually tried both approaches (binary cross entropy + sigmoid (2 classes) and cross entropy+softmax (3 classes). And essentially it's very similar because in the latter, the paper states to weight the background as zero. I actually tried to do the post-processing part of segment merging but it's not THAT easy. Also the Matlab code of the segment merging was confusing to me so I'm in the process of doing it another way more simple way (maybe less accurate).  In terms of training, I actually used a private dataset, but the leaves were also dense. Hopefully, it will be open-sourced in the future.

I trained on 384x384 resolution stacking the pyramid 4 times, so from 384/192/96/48. Problem in training was at first that the overall loss was stuck at around 2.5. Then after applying extensive data augmentation (before training rather than applying augmentations on the go during training), it went down to 1.4. However, the problem persists and I reckon this can be still optimized. 

Have you thought of implementing the loss differently than summing all the outputs? At first I thought of maybe weighting the losses for each resolution output or something similar.

**Comment by NickLucche** [comment]
Thanks again for the extensive report! 
Feel free to share any additional updates regarding the post-processing or dataset too.
 
I had a similar problem reading the few Matlab lines I needed..

Okay so you trained at a higher resolution, have you considered giving more "blocks" to each pyramid layer? I think I stuck with all the hyperparameters the paper proposed, I believe the number of blocks per-layer was something like 4-5, but I didn't have much time to experiment with different settings.

Yes giving different weights to each loss is surely a vaiable option, specifically for this application lower resolutions converge really quickly, as it is to be expected, so it would be worth trying having increasingly higher weights for each loss as the resolution goes up.
